### ISSUE TITLE
`refs/heads/<branchName>` spec in Bitbucket server push action seems not working properly

### ISSUE THREAD

**Issue by hexonxons** [jenkinsci/bitbucket-push-and-pull-request-plugin#149]
I'm creating job to run on push to `dev` branch in Bitbucket git repo. According to the documentation I need to specify `Allowed Branches` in `Push` action. So I did:
![image](https://user-images.githubusercontent.com/663637/99959645-7110b300-2d9c-11eb-9366-dc26161ca78c.png)

Also it is written `Better use refs/heads/<branchName>` instead of just branch name, so I have written: `refs/heads/dev`. And nothing worked.

So I tried to investigate what happened.
<details>
  <summary>Jenkins log</summary>

```
Nov 23, 2020 12:01:29 PM FINE io.jenkins.plugins.bitbucketpushandpullrequest.BitBucketPPRHookReceiver
Received commit hook notification : {"eventKey":"repo:refs_changed","date":"2020-11-23T12:01:29+0000","actor":{"name":"hexonxons","emailAddress":"hexonxons@gmail.com","id":1,"displayName":"hexonxons","active":true,"slug":"hexonxons","type":"NORMAL","links":{"self":[{"href":"https://bitbucket.hexonxons.me/users/hexonxons"}]}},"repository":{"slug":"build","id":2,"name":"build","hierarchyId":"8cbad0a465844d48664e","scmId":"git","state":"AVAILABLE","statusMessage":"Available","forkable":true,"project":{"key":"INFRA","id":1,"name":"Infra","public":false,"type":"NORMAL","links":{"self":[{"href":"https://bitbucket.hexonxons.me/projects/INFRA"}]}},"public":false,"links":{"clone":[{"href":"https://bitbucket.hexonxons.me/scm/infra/build.git","name":"http"},{"href":"ssh://git@bitbucket.hexonxons.me:7999/infra/build.git","name":"ssh"}],"self":[{"href":"https://bitbucket.hexonxons.me/projects/INFRA/repos/build/browse"}]}},"changes":[{"ref":{"id":"refs/heads/dev","displayId":"dev","type":"BRANCH"},"refId":"refs/heads/dev","fromHash":"8a931e9a3fcc025caf15ed1dfbaec342d1a0ef6f","toHash":"6e1ffd142d5bf7a0460c033b3d117e8ce84f0b66","type":"UPDATE"}]}
Nov 23, 2020 12:01:29 PM INFO io.jenkins.plugins.bitbucketpushandpullrequest.BitBucketPPRHookReceiver doIndex
Received x-event-key repo:refs_changed from Bitbucket
Nov 23, 2020 12:01:29 PM FINEST io.jenkins.plugins.bitbucketpushandpullrequest.BitBucketPPRHookReceiver
the payload is: {"actor":{"name":"hexonxons","emailAddress":"hexonxons@gmail.com","id":"1","displayName":"hexonxons","active":true,"slug":"hexonxons","type":"NORMAL","links":{"clone":[],"self":[{"href":"https://bitbucket.hexonxons.me/users/hexonxons"}]}},"repository":{"slug":"build","id":"2","name":"build","scmId":"git","state":"AVAILABLE","statusMessage":"Available","forkable":true,"project":{"key":"INFRA","id":"1","name":"Infra","links":{"clone":[],"self":[{"href":"https://bitbucket.hexonxons.me/projects/INFRA"}]},"public":false,"type":"NORMAL"},"links":{"clone":[{"href":"https://bitbucket.hexonxons.me/scm/infra/build.git","name":"http"},{"href":"ssh://git@bitbucket.hexonxons.me:7999/infra/build.git","name":"ssh"}],"self":[{"href":"https://bitbucket.hexonxons.me/projects/INFRA/repos/build/browse"}]},"public":false},"changes":[{"ref":{"id":"refs/heads/dev","displayId":"dev","type":"BRANCH"},"refId":"refs/heads/dev","fromHash":"8a931e9a3fcc025caf15ed1dfbaec342d1a0ef6f","toHash":"6e1ffd142d5bf7a0460c033b3d117e8ce84f0b66","type":"UPDATE"}]}
Nov 23, 2020 12:01:29 PM INFO io.jenkins.plugins.bitbucketpushandpullrequest.processor.BitBucketPPRPayloadProcessorFactory createProcessor
Create BitBucketPPRRepositoryServerPayloadProcessor
Nov 23, 2020 12:01:29 PM INFO io.jenkins.plugins.bitbucketpushandpullrequest.observer.BitBucketPPRObserverFactory createObservable
Add BitBucketPPRPushServerObserver for BitBucketPPREvent [event=repo, action=refs_changed]
Nov 23, 2020 12:01:29 PM FINEST io.jenkins.plugins.bitbucketpushandpullrequest.BitBucketPPRHookReceiver
the selected payload processor is: io.jenkins.plugins.bitbucketpushandpullrequest.processor.BitBucketPPRRepositoryServerPayloadProcessor@4a923ed8
Nov 23, 2020 12:01:29 PM INFO io.jenkins.plugins.bitbucketpushandpullrequest.action.BitBucketPPRServerRepositoryAction 
Received commit hook notification from server for destination branch: dev
Nov 23, 2020 12:01:29 PM INFO io.jenkins.plugins.bitbucketpushandpullrequest.action.BitBucketPPRServerRepositoryAction 
Received commit hook type from server: BRANCH
Nov 23, 2020 12:01:29 PM FINE io.jenkins.plugins.bitbucketpushandpullrequest.BitBucketPPRJobProbe
Considering remote [https://bitbucket.hexonxons.me/scm/infra/build.git, ssh://git@bitbucket.hexonxons.me:7999/infra/build.git]
Nov 23, 2020 12:01:29 PM FINE io.jenkins.plugins.bitbucketpushandpullrequest.BitBucketPPRJobProbe
Considering candidate job Push
Nov 23, 2020 12:01:29 PM FINE io.jenkins.plugins.bitbucketpushandpullrequest.BitBucketPPRJobProbe
Considering to poke bbppr » build » Push
Nov 23, 2020 12:01:29 PM FINE io.jenkins.plugins.bitbucketpushandpullrequest.BitBucketPPRJobProbe
Considering to use trigger hudson.plugins.git.GitSCM@301107c1
Nov 23, 2020 12:01:29 PM FINE io.jenkins.plugins.bitbucketpushandpullrequest.BitBucketPPRJobProbe
Job is of type: com.cloudbees.hudson.plugins.folder.Folder
Nov 23, 2020 12:01:29 PM FINE io.jenkins.plugins.bitbucketpushandpullrequest.BitBucketPPRJobProbe
Scheduling for job:Push
Nov 23, 2020 12:01:29 PM INFO io.jenkins.plugins.bitbucketpushandpullrequest.BitBucketPPRJobProbe matchGitScm
Trying to match ssh://git@bitbucket.hexonxons.me:7999/infra/build.git<-->https://bitbucket.hexonxons.me/scm/infra/build.git 
Nov 23, 2020 12:01:29 PM INFO io.jenkins.plugins.bitbucketpushandpullrequest.BitBucketPPRJobProbe matchGitScm
Trying to match ssh://git@bitbucket.hexonxons.me:7999/infra/build.git<-->ssh://git@bitbucket.hexonxons.me:7999/infra/build.git 
Nov 23, 2020 12:01:29 PM INFO io.jenkins.plugins.bitbucketpushandpullrequest.BitBucketPPRJobProbe matchGitScm
Matched scm
Nov 23, 2020 12:01:29 PM FINE io.jenkins.plugins.bitbucketpushandpullrequest.BitBucketPPRJobProbe
Triggering trigger io.jenkins.plugins.bitbucketpushandpullrequest.BitBucketPPRTrigger for job bbppr » build » Push
Nov 23, 2020 12:01:29 PM INFO io.jenkins.plugins.bitbucketpushandpullrequest.BitBucketPPRTrigger onPost
Called onPost with BitBucketPPRServerRepositoryAction
Nov 23, 2020 12:01:29 PM INFO io.jenkins.plugins.bitbucketpushandpullrequest.filter.repository.BitBucketPPRRepositoryTriggerMatcher matchesAction
BitBucketPPREvent [event=repo, action=refs_changed]
Nov 23, 2020 12:01:29 PM FINE io.jenkins.plugins.bitbucketpushandpullrequest.filter.repository.BitBucketPPRRepositoryTriggerMatcher
1. (Is the trigger filter instance of BitBucketPPRRepositoryPushActionFilter? <<false>> AND does it equal BitBucketPPRConsts.REPOSITORY_PUSH push to bitbucketEvent.getAction() refs_changed? <<false>>) OR 2. (Is the trigger filter instance of BitBucketPPRServerRepositoryPushActionFilter? <<true>>AND does it equal BitBucketPPRConsts.REPOSITORY_SERVER_PUSH refs_changed to bitbucketEvent.getAction() refs_changed? <<true>> 
Nov 23, 2020 12:01:29 PM INFO io.jenkins.plugins.bitbucketpushandpullrequest.filter.BitBucketPPRFilterMatcher getMatchingFilters
Event BitBucketPPREvent [event=repo, action=refs_changed] matches trigger filter io.jenkins.plugins.bitbucketpushandpullrequest.filter.repository.BitBucketPPRRepositoryTriggerFilter@391cd863
Nov 23, 2020 12:01:29 PM INFO io.jenkins.plugins.bitbucketpushandpullrequest.BitBucketPPRTrigger onPost
matchingFilters is not null AND is not empty [io.jenkins.plugins.bitbucketpushandpullrequest.filter.repository.BitBucketPPRRepositoryTriggerFilter@391cd863] 
Nov 23, 2020 12:01:29 PM FINE io.jenkins.plugins.bitbucketpushandpullrequest.BitBucketPPRJobProbe
Considering to use trigger hudson.plugins.git.GitSCM@5b655351
Nov 23, 2020 12:01:29 PM FINE io.jenkins.plugins.bitbucketpushandpullrequest.BitBucketPPRJobProbe
Job is of type: com.cloudbees.hudson.plugins.folder.Folder
Nov 23, 2020 12:01:29 PM FINE io.jenkins.plugins.bitbucketpushandpullrequest.BitBucketPPRJobProbe
Scheduling for job:Push
Nov 23, 2020 12:01:29 PM INFO io.jenkins.plugins.bitbucketpushandpullrequest.BitBucketPPRJobProbe matchGitScm
Trying to match ssh://git@bitbucket.hexonxons.me:7999/infra/devops.git<-->https://bitbucket.hexonxons.me/scm/infra/build.git 
Nov 23, 2020 12:01:29 PM FINE io.jenkins.plugins.bitbucketpushandpullrequest.BitBucketPPRPollingRunnable
Run method called.
Nov 23, 2020 12:01:29 PM INFO io.jenkins.plugins.bitbucketpushandpullrequest.BitBucketPPRJobProbe matchGitScm
Trying to match ssh://git@bitbucket.hexonxons.me:7999/infra/devops.git<-->ssh://git@bitbucket.hexonxons.me:7999/infra/build.git 
Nov 23, 2020 12:01:29 PM FINE io.jenkins.plugins.bitbucketpushandpullrequest.BitBucketPPRJobProbe
Push SCM doesnt match remote repo {1}
Nov 23, 2020 12:01:29 PM FINEST io.jenkins.plugins.bitbucketpushandpullrequest.BitBucketPPRHookReceiver
Sending response.
Nov 23, 2020 12:01:29 PM INFO io.jenkins.plugins.bitbucketpushandpullrequest.BitBucketPPRHookReceiver doIndex
Response sent.
Nov 23, 2020 12:01:30 PM INFO io.jenkins.plugins.bitbucketpushandpullrequest.BitBucketPPRTrigger$1 onPollSuccess
Called onPollSuccess with polling result hudson.scm.PollingResult@4e96f1ed
Nov 23, 2020 12:01:30 PM INFO io.jenkins.plugins.bitbucketpushandpullrequest.BitBucketPPRTrigger$1 onPollSuccess
On Poll Success, get cause: BitBucketPPRTriggerCause [bitbucketAction=BitBucketPPRServerRepositoryAction]
Nov 23, 2020 12:01:30 PM INFO io.jenkins.plugins.bitbucketpushandpullrequest.filter.repository.BitBucketPPRServerRepositoryPushActionFilter shouldTriggerBuild
Should trigger build for bitbucket actionBitBucketPPRServerRepositoryAction?
Nov 23, 2020 12:01:30 PM FINEST io.jenkins.plugins.bitbucketpushandpullrequest.filter.repository.BitBucketPPRServerRepositoryPushActionFilter
the target branch is: dev.
Nov 23, 2020 12:01:30 PM FINEST io.jenkins.plugins.bitbucketpushandpullrequest.filter.repository.BitBucketPPRServerRepositoryPushActionFilter
The allowed branches are: refs/heads/dev.
Nov 23, 2020 12:01:30 PM FINEST io.jenkins.plugins.bitbucketpushandpullrequest.util.BitBucketPPRUtils
Following allowed branches patterns are set: refs/heads/dev
Nov 23, 2020 12:01:30 PM FINEST io.jenkins.plugins.bitbucketpushandpullrequest.util.BitBucketPPRUtils
The branchName in action is: dev
Nov 23, 2020 12:01:30 PM FINEST io.jenkins.plugins.bitbucketpushandpullrequest.util.BitBucketPPRUtils
The environment variables are: null
Nov 23, 2020 12:01:30 PM INFO io.jenkins.plugins.bitbucketpushandpullrequest.BitBucketPPRTrigger shouldScheduleJob
Should schedule job: false and (polling result has changes: true or trigger also if there arent changes: {2})
Nov 23, 2020 12:01:30 PM INFO io.jenkins.plugins.bitbucketpushandpullrequest.filter.repository.BitBucketPPRServerRepositoryPushActionFilter shouldTriggerBuild
Should trigger build for bitbucket actionBitBucketPPRServerRepositoryAction?
Nov 23, 2020 12:01:30 PM FINEST io.jenkins.plugins.bitbucketpushandpullrequest.filter.repository.BitBucketPPRServerRepositoryPushActionFilter
the target branch is: dev.
Nov 23, 2020 12:01:30 PM FINEST io.jenkins.plugins.bitbucketpushandpullrequest.filter.repository.BitBucketPPRServerRepositoryPushActionFilter
The allowed branches are: refs/heads/dev.
Nov 23, 2020 12:01:30 PM FINEST io.jenkins.plugins.bitbucketpushandpullrequest.util.BitBucketPPRUtils
Following allowed branches patterns are set: refs/heads/dev
Nov 23, 2020 12:01:30 PM FINEST io.jenkins.plugins.bitbucketpushandpullrequest.util.BitBucketPPRUtils
The branchName in action is: dev
Nov 23, 2020 12:01:30 PM FINEST io.jenkins.plugins.bitbucketpushandpullrequest.util.BitBucketPPRUtils
The environment variables are: null
```

</details>
Trying search by log messages and found next logic:

https://github.com/jenkinsci/bitbucket-push-and-pull-request-plugin/blob/fdf08f939e128290492e90f2a6d083450a8e62ac/src/main/java/io/jenkins/plugins/bitbucketpushandpullrequest/filter/repository/BitBucketPPRServerRepositoryPushActionFilter.java#L90-L100

`target` variable is set to `bitbucketAction.getTargetBranch()`, which is `change.getRef().getDisplayId()`:
https://github.com/jenkinsci/bitbucket-push-and-pull-request-plugin/blob/fdf08f939e128290492e90f2a6d083450a8e62ac/src/main/java/io/jenkins/plugins/bitbucketpushandpullrequest/action/BitBucketPPRServerRepositoryAction.java#L60-L66

Which is `dev` according to given Jenkins log. And call
https://github.com/jenkinsci/bitbucket-push-and-pull-request-plugin/blob/fdf08f939e128290492e90f2a6d083450a8e62ac/src/main/java/io/jenkins/plugins/bitbucketpushandpullrequest/filter/repository/BitBucketPPRServerRepositoryPushActionFilter.java#L99

fails, cuz of `dev` is not matched by `refs/heads/dev`.

Seems like there is a bug in the code(wrong logic in selecting target branch name) or bug in the documentation(don't think so, but still).

**Comment by cdelmonte-zg** [comment]
@hexonxons how is written the target branch in the  payload BB is sending to Jenkins?

**Comment by hexonxons** [comment]
@cdelmonte-zg 
In jenkins log under expand:
```
{"actor":{"name":"hexonxons","emailAddress":"hexonxons@gmail.com","id":"1","displayName":"hexonxons","active":true,"slug":"hexonxons","type":"NORMAL","links":{"clone":[],"self":[{"href":"https://bitbucket.hexonxons.me/users/hexonxons"}]}},"repository":{"slug":"build","id":"2","name":"build","scmId":"git","state":"AVAILABLE","statusMessage":"Available","forkable":true,"project":{"key":"INFRA","id":"1","name":"Infra","links":{"clone":[],"self":[{"href":"https://bitbucket.hexonxons.me/projects/INFRA"}]},"public":false,"type":"NORMAL"},"links":{"clone":[{"href":"https://bitbucket.hexonxons.me/scm/infra/build.git","name":"http"},{"href":"ssh://git@bitbucket.hexonxons.me:7999/infra/build.git","name":"ssh"}],"self":[{"href":"https://bitbucket.hexonxons.me/projects/INFRA/repos/build/browse"}]},"public":false},"changes":[{"ref":{"id":"refs/heads/dev","displayId":"dev","type":"BRANCH"},"refId":"refs/heads/dev","fromHash":"8a931e9a3fcc025caf15ed1dfbaec342d1a0ef6f","toHash":"6e1ffd142d5bf7a0460c033b3d117e8ce84f0b66","type":"UPDATE"}]}
```
`changes[0].ref.id = "refs/heads/dev"`, `changes[0].ref.displayId = "dev"`

**Comment by cdelmonte-zg** [comment]
@hexonxons should be fixed in v2.6.1, released just now

**Comment by hexonxons** [comment]
Yep, now it works properly.
Thank you!
